sdk/program/src/vote/state/mod.rs

```diff
@@ -479,13 +479,43 @@ impl VoteState {
         }
     }
 
-    /// Deserializes the input `VoteStateVersions` buffer directly into a provided `VoteState` struct
+    /// Deserializes the input `VoteStateVersions` buffer directly into the provided `VoteState`.
     ///
-    /// In a BPF context, V0_23_5 is not supported, but in non-BPF, all versions are supported for
-    /// compatibility with `bincode::deserialize`
+    /// In a SBPF context, V0_23_5 is not supported, but in non-SBPF, all versions are supported for
+    /// compatibility with `bincode::deserialize`.
     pub fn deserialize_into(
         input: &[u8],
         vote_state: &mut VoteState,
+    ) -> Result<(), InstructionError> {
+        // Safety: vote_state is valid to_drop (see drop_in_place() docs). After
+        // dropping, the pointer is treated as uninitialized and only accessed
+        // through ptr::write, which is safe as per drop_in_place docs.
+        unsafe {
+            std::ptr::drop_in_place(vote_state);
+        }
+
+        VoteState::deserialize_into_ptr(input, vote_state)
+    }
+
+    /// Deserializes the input `VoteStateVersions` buffer directly into the provided
+    /// `MaybeUninit<VoteState>`.
+    ///
+    /// In a SBPF context, V0_23_5 is not supported, but in non-SBPF, all versions are supported for
+    /// compatibility with `bincode::deserialize`.
+    ///
+    /// On success, `vote_state` is fully initialized and can be converted to `VoteState` using
+    /// [MaybeUninit::assume_init]. On failure, `vote_state` may still be uninitialized and must not
+    /// be converted to `VoteState`.
+    pub fn deserialize_into_uninit(
+        input: &[u8],
+        vote_state: &mut MaybeUninit<VoteState>,
+    ) -> Result<(), InstructionError> {
+        VoteState::deserialize_into_ptr(input, vote_state.as_mut_ptr())
+    }
```

Comment:
Good question!

The point of this function is to deserialize into a _provided_ `VoteState` and not create a new `VoteState`. What you are suggesting always creates a new value (`let mut vote_state = MaybeUninit::uninit()`). 

There are two reasons for wanting to deserialize into a provided value:

- take `VoteAccount`, where `VoteState` is a subfield of `VoteAccount` - by deserializing into a given `VoteState` I can deserialize in place, skipping a memcpy.
- what you're suggesting is similar to what `serde::Deserialize::deserialize` does. Because LLVM/rustc aren't great at NRVO, even if you do `let state = VoteState::deserialize(input)` often the compiler won't create the result straight in the caller frame, but it will create it in the callee frame and then memcpy it back to the caller. In a followup PR, I show how this is currently happening when we do `Arc::new(vote_account)` inside `VoteAccount::try_from(data)`, where rustc first creates `vote_account` then memcpys it into the `Arc` allocation because computers were a mistake and everything is terrible

Reply:
>  will create it in the callee frame and then memcpy it back to the caller
should mention that this is especially bad for `VoteState` because `VoteState::prior_voters` is a 1k byte array so memcpy cost is not negligible

Reply:
> Ahh, that's the context I didn't have. So the real caller will have an uninitialized VoteAccount, and then call deserialize_into_uninit()? And that would be to avoid having to call deserialize_into_ptr() directly then?

correct! VoteAccount is in another crate. I don't want to expose `deserialize_into_ptr` and pointers are gross anyway, `MaybeUninit` is what rust provides to do what we want to do here.